### PR TITLE
docs: add missing document to microsoft.extensions.options

### PR DESCRIPTION
**Pull request by ikesnowy**
I couldn't find some of the apis, maybe they are no longer valid, or I missed something.

- ``M:Microsoft.Extensions.Options.OptionsWrapper`1.Add(System.String,`0)`` (No such method in `OptionsWrapper`)
- ``M:Microsoft.Extensions.Options.OptionsWrapper`1.Remove(System.String)`` (No such method in `OptionsWrapper`)
- ``M:Microsoft.Extensions.Options.OptionsCache`1.#ctor`` (``OptionsCache`` uses default constructor)
- ``M:Microsoft.Extensions.Options.ValidateOptionsResult.#ctor`` (``ValidateOptionsResult`` uses default constructor)
- `T:Microsoft.Extensions.Options.OptionsSnapshot` (No such type)

Contributes to https://github.com/dotnet/runtime/issues/43919

### CODE REVIEW

**Comment by ikesnowy** [comment]
@carlossanlop asked about reviewer but I don't have the permission.

File: src/libraries/Microsoft.Extensions.Options/src/IConfigureNamedOptions.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.Extensions.Options
     /// <summary>
     /// Represents something that configures the <typeparamref name="TOptions"/> type.
     /// </summary>
-    /// <typeparam name="TOptions"></typeparam>
+    /// <typeparam name="TOptions">Options type being configured.</typeparam>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IConfigureNamedOptions.cs:9]:
```suggestion
    /// <typeparam name="TOptions">The options type being configured.</typeparam>
```

File: src/libraries/Microsoft.Extensions.Options/src/IConfigureOptions.cs [hunk 2]
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Extensions.Options
     /// Represents something that configures the <typeparamref name="TOptions"/> type.
     /// Note: These are run before all <see cref="IPostConfigureOptions{TOptions}"/>.
     /// </summary>
-    /// <typeparam name="TOptions"></typeparam>
+    /// <typeparam name="TOptions">Options type being configuring.</typeparam>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IConfigureOptions.cs:10]:
```suggestion
    /// <typeparam name="TOptions">The options type being configuring.</typeparam>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsChangeTokenSource.cs [hunk 3]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.Extensions.Options
     /// <summary>
     /// Used to fetch <see cref="IChangeToken"/> used for tracking options changes.
     /// </summary>
-    /// <typeparam name="TOptions">Options type.</typeparam>
+    /// <typeparam name="TOptions">Options type being changed.</typeparam>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsChangeTokenSource.cs:11]:
```suggestion
    /// <typeparam name="TOptions">The options type being changed.</typeparam>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs [hunk 4]
```diff
@@ -13,8 +14,12 @@ public interface IOptionsFactory<[DynamicallyAccessedMembers(Options.Dynamically
         where TOptions : class
     {
         /// <summary>
-        /// Returns a configured <typeparamref name="TOptions"/> instance with the given name.
+        /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of <typeparamref name="TOptions"/> instance to create.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs:19]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance to create.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsMonitor.cs [hunk 5]
```diff
@@ -18,8 +18,12 @@ public interface IOptionsMonitor<[DynamicallyAccessedMembers(Options.Dynamically
         TOptions CurrentValue { get; }
 
         /// <summary>
-        /// Returns a configured <typeparamref name="TOptions"/> instance with the given name.
+        /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if null <see cref="Options.DefaultName"/> is used.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsMonitor.cs:23]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if a <see langword="null"/> <see cref="Options.DefaultName"/> is used.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsMonitor.cs [hunk 6]
```diff
@@ -18,8 +18,12 @@ public interface IOptionsMonitor<[DynamicallyAccessedMembers(Options.Dynamically
...
+        /// <returns>The <typeparamref name="TOptions"/> instances with given <paramref name="name"/>.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsMonitor.cs:24]:
```suggestion
        /// <returns>The <typeparamref name="TOptions"/> instance that matches the given <paramref name="name"/>.</returns>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsSnapshot.cs [hunk 7]
```diff
@@ -14,8 +15,12 @@ public interface IOptionsSnapshot<[DynamicallyAccessedMembers(Options.Dynamicall
         where TOptions : class
     {
         /// <summary>
-        /// Returns a configured <typeparamref name="TOptions"/> instance with the given name.
+        /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if null <see cref="Options.DefaultName"/> is used.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsSnapshot.cs:20]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if <see langword="null"/> <see cref="Options.DefaultName"/> is used.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsSnapshot.cs [hunk 8]
```diff
@@ -14,8 +15,12 @@ public interface IOptionsSnapshot<[DynamicallyAccessedMembers(Options.Dynamicall
...
+        /// <returns>The <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsSnapshot.cs:21]:
```suggestion
        /// <returns>The <typeparamref name="TOptions"/> instance that matches the given <paramref name="name"/>.</returns>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 9]
```diff
@@ -48,6 +48,10 @@ public OptionsFactory(IEnumerable<IConfigureOptions<TOptions>> setups, IEnumerab
         /// <summary>
         /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of <typeparamref name="TOptions"/> instance to create.</param>
+        /// <returns>The created <typeparamref name="TOptions"/> instance with thw given <paramref name="name"/>.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:52]:
```suggestion
        /// <returns>The created <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.</returns>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 10]
```diff
@@ -48,6 +48,10 @@ public OptionsFactory(IEnumerable<IConfigureOptions<TOptions>> setups, IEnumerab
         /// <summary>
         /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of <typeparamref name="TOptions"/> instance to create.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:51]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance to create.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 11]
```diff
@@ -90,6 +94,9 @@ public TOptions Create(string name)
         /// <summary>
         /// Creates a new instance of options type
         /// </summary>
+        /// <param name="name">The name of <typeparamref name="TOptions"/> instance.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:97]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance to create.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 12]
```diff
@@ -90,6 +94,9 @@ public TOptions Create(string name)
         /// <summary>
         /// Creates a new instance of options type
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:95]: Not part of your PR but needs a tiny fix:
```suggestion
        /// Creates a new instance of options type.
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 13]
```diff
@@ -90,6 +94,9 @@ public TOptions Create(string name)
...
         /// </summary>
+        /// <param name="name">The name of <typeparamref name="TOptions"/> instance.</param>
+        /// <returns>The <typeparamref name="TOptions"/> instance been created.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:98]:
```suggestion
        /// <returns>The created <typeparamref name="TOptions"/> instance.</returns>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsManager.cs [hunk 14]
```diff
@@ -34,6 +35,10 @@ public OptionsManager(IOptionsFactory<TOptions> factory)
         /// <summary>
         /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if null <see cref="Options.DefaultName"/> is used.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsManager.cs:38]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if <see langword="null"/> <see cref="Options.DefaultName"/> is used.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsManager.cs [hunk 15]
```diff
@@ -34,6 +35,10 @@ public OptionsManager(IOptionsFactory<TOptions> factory)
...
+        /// <returns>The <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsManager.cs:39]:
```suggestion
        /// <returns>The <typeparamref name="TOptions"/> instance that matches the given <paramref name="name"/>.</returns>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsMonitor.cs [hunk 16]
```diff
@@ -80,6 +82,10 @@ public TOptions CurrentValue
         /// <summary>
         /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
+        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if null <see cref="Options.DefaultName"/> is used.</param>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsMonitor.cs:85]:
```suggestion
        /// <param name="name">The name of the <typeparamref name="TOptions"/> instance, if <see langword="null"/> <see cref="Options.DefaultName"/> is used.</param>
```

File: src/libraries/Microsoft.Extensions.Options/src/OptionsMonitor.cs [hunk 17]
```diff
@@ -80,6 +82,10 @@ public TOptions CurrentValue
...
+        /// <returns>The <typeparamref name="TOptions"/> instances with given <paramref name="name"/>.</returns>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsMonitor.cs:86]:
```suggestion
        /// <returns>The <typeparamref name="TOptions"/> instance that matches the given <paramref name="name"/>.</returns>
```

**Review by carlossanlop — COMMENTED** [review]
Thanks for your contribution! Left a few suggestions for you to consider.

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs [hunk 18]
```diff
@@ -13,8 +14,12 @@ public interface IOptionsFactory<[DynamicallyAccessedMembers(Options.Dynamically
...
+        /// <returns>The created <typeparamref name="TOptions"/> instance with thw given <paramref name="name"/>.</returns>
+        /// <exception cref="OptionsValidationException">One or more <see cref="IValidateOptions{TOptions}"/> return failed <see cref="ValidateOptionsResult"/> when validating the <typeparamref name="TOptions"/> instance been created.</exception>
+        /// <exception cref="MissingMethodException">The <typeparamref name="TOptions"/> does not have a public parameterless constructor or <typeparamref name="TOptions"/> is <see langword="abstract"/>.</exception>
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs:22]: Just wondering, do we mention the exceptions in the interface docs? I think the exception is mainly thrown from the actual interface implementation.

**Comment by tarekgh** [comment]
@ikesnowy thanks for teh PR. 

I added a little comment/question. Other than @carlossanlop comments, LGTM.

File: src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs (hunk 18 above)
Comment (carlossanlop) [reply src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs:22]: That is a very good question, I don't think we should have exceptions here.

Comment (ikesnowy) [reply src/libraries/Microsoft.Extensions.Options/src/IOptionsFactory.cs]: Done, I removed the exceptions from interface docs.

**Review by tarekgh — APPROVED** [review]

**Comment by tarekgh** [comment]
@carlossanlop if you are ok with the changes we can merge this one.

**Pull request merged**